src/transformStateWithClones.js

```diff
@@ -7,7 +7,25 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  actions.forEach((action) => {
+    if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    } else if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      action.keysToRemove.forEach((key) => {
+        delete currentState[key];
+      });
+    } else if (action.type === 'clear') {
+      currentState = {};
+    }
```

Comment:
Consider using a switch statement instead of multiple if-else conditions for better readability and maintainability, especially since you have a limited and known set of action types.